README.md
# pyjbot


A test bot for Facebook Messenger and Wit.ai hosted on heroku. 

Still gltichy and takes a while to start up but there is some progress.


CHANGES.md
#To-Do List:
- Some form of database for recipes
- Consolidate recipe ingredients for multiple recipes
- Other Wit story types
- Search function
- Song sender
- News filter
- Praw/reddit module?
- Help option
- Google places other requests? Maybe keyword search

#V6.7
- Added google places api support
- Added "What restaurants?" request
- Can recieve and store locaiton data for at least 30 min?

#V6.6
- Splits up larger messages to fit messenger char limit
- Working on search function
- Checks if recieved messages are part of story sets, otherwise asks for a rephrase

#V6.5
- Eliza.py added
- Basically just reflects questions, not bad
- Working on weather request next

#V6.4
- Memory leak issues with aiml bot.
- Trying a different chat bot approach later
- Implementing weather

#V6.3
- Attempting to add A.L.I.C.E to the fb bot
- Need to find a way to loop responses inside event handling for the aiml bot
- Solved above, needed to set a flag to get around normal event handling

#V6.2
- Scrapped the Wit.ai run_actions section
- 500 errors from somewhere
- Fixed infinite hi
- FB messages are limited to 320 chars
- Limited to ingredients of one recipe
- Recipe ingredient requests now work
- Lxml since python parser is not working for bs4

#V6.1
- Facebook support now works (It can connect and send/recieve messages)
- Infinite loop on greetings from wit.ai
- Page chat only so the bot is stuck with single chat
- Basis for the bot server came from hult's flask-python chatbot

#V6
- Removing pygooglevoice
- Attempting to use flask to create a fb messenger bot
- Removed weather command support 

#V5
- Can use wit to determine what type of response is required
- Can run methods from recieved messages
- Response times vary and pygooglevoice encounters parsing errors


p.py
from flask import Flask, request
import json
import requests

from wit import Wit
import BeautifulSoup
import time

from bstest6_3 import foodSites


app = Flask(__name__)

# This needs to be filled with the Page Access Token that will be provided
# by the Facebook App that will be created.
PAT = 'EAAXQgdDXC2kBAI2zLHJSZA6Aapz1iPCodArxZAU2lDZAmt7IYrKgewl6a6OWvsCRx0nv6hvhrAiIPmhUKBffeJ9V7YGDoZBENixzZCPqpEM3OvZBcBdZCori4RtBC5nVdJZA6mfLgoIOruJE96Xvo0ZBqIPu6OXzBwbCxVzOE8mcoyAZDZD'
@app.route('/', methods=['GET'])
def handle_verification():
  print "Handling Verification."
  if request.args.get('hub.verify_token', '') == 'my_voice_is_my_password_verify_me':
    print "Verification successful!"
    return request.args.get('hub.challenge', '')
  else:
    print "Verification failed!"
    return 'Error, wrong validation token'

@app.route('/', methods=['POST'])
def handle_messages():
  context0 = {}
  print "Handling Messages"
  payload = request.get_data()
  print payload
  for sender, message in messaging_events(payload):
    print "Incoming from %s: %s" % (sender, message)
    print type(message)
    resp = client.message(message)
    resp = resp[u'entities']
    resp = resp[u'intent']
    resp = resp[0]
    print ("Response type is.... "+resp[u'value'])
    if u'value' in resp:
      if resp[u'value'] == "grocery":
        print("THIS GROCERY RIGHT HERE")
        resp = client.run_actions('my-user-session-42',message, context0)
        print("This resp grocery ")
        print (resp)
        while('foodList' not in resp):
          resp = client.run_actions('my-user-session-42',message, context0)
          print ("This resp ")
          print(resp)
        message = str(resp['foodList'])
        send_message(PAT, sender, message)
         
      elif resp[u'value'] == "greetings":
        print("This resp greetings RIGHT HERE")
        resp = client.converse('my-user-session-42',message, context0)
        print("This resp greetings ")
        print (resp)
        while('msg' not in resp):
          resp = client.converse('my-user-session-42',message, context0)
          print ("This resp HERE ")
          print(resp)
          
        print("the msg is "+resp['msg'])
        message = str(resp["msg"])
        print("Trying to send...")
        send_message(PAT, sender, message)
        print("Probably sent")
      elif resp[u'value'] == "weather":
        #resp = client.run_actions('my-user-session-42',textmsg, context0)
        print("This resp weather ")
        #print (resp)
        #while('foodList' not in resp):
        #    resp = client.run_actions('my-user-session-42',textmsg, context0)
        #    print ("This resp ")
        #    print(resp)
        #voice.send_sms(msg[u'from'],str(resp['forecast']))
        message = "weather"
        send_message(PAT, sender, message)
      else:
        print("Else")
        message = "else"
        send_message(PAT, sender, message)
  return "ok"

def messaging_events(payload):
  """Generate tuples of (sender_id, message_text) from the
  provided payload.
  """
  data = json.loads(payload)
  messaging_events = data["entry"][0]["messaging"]
  for event in messaging_events:
    if "message" in event and "text" in event["message"]:
      yield event["sender"]["id"], event["message"]["text"].encode('unicode_escape')
    else:
      yield event["sender"]["id"], "I can't echo this"


def send_message(token, recipient, text):
  """Send the message text to recipient with id recipient.
  """

  r = requests.post("https://graph.facebook.com/v2.6/me/messages",
    params={"access_token": token},
    data=json.dumps({
      "recipient": {"id": recipient},
      "message": {"text": text.decode('unicode_escape')}
    }),
    headers={'Content-type': 'application/json'})
  if r.status_code != requests.codes.ok:
    print r.text

def merge():
    return 0

def error():
    return 0

def say(id, dict, response):
    print ("id " +id)
    print (dict)
    print ("respo "+response)
    return response
    
def send(request, response):
    print(response['text'])

def my_action(request):
    print('Received from user...', request['text'])

def get_forecast(request):
    context = request['context']
    entities = request['entities']

    loc = first_entity_value(entities, 'location')
    if loc:
        context['forecast'] = 'sunny'
    else:
        context['missingLocation'] = True
        if context.get('forecast') is not None:
            del context['forecast']

    return context

def get_cooking(ent, cont):
    print("Inside grocery")
    print(cont)
    print (ent)
    context = cont
    cook = foodSites()
    cook.initList()
    context['foodList'] = cook.getIngred()
    return context
    

actions = {
    'send': send,
    'getForecast': get_forecast,
    'say': say,
    'merge': merge,
    'error': error,
    "getCooking": get_cooking,
    'my_action': my_action,
}

#server access
client = Wit(access_token="ZNJWKAFJBJI4UXBMTL2O4YW2RNKOSABS", actions=actions)

#client token
#client = Wit(access_token="2F5GRMKTGCRTVHIAZO7HXY64LFFSVYWL", actions=actions)
#client.interactive()


if __name__ == '__main__':
  app.run()


eilza.py
import re
import random
 
 
reflections = {
    "am": "are",
    "was": "were",
    "i": "you",
    "i'd": "you would",
    "i've": "you have",
    "i'll": "you will",
    "my": "your",
    "are": "am",
    "you've": "I have",
    "you'll": "I will",
    "your": "my",
    "yours": "mine",
    "you": "me",
    "me": "you"
}
 
psychobabble = [
    [r'I need (.*)',
     ["Why do you need {0}?",
      "Would it really help you to get {0}?",
      "Are you sure you need {0}?"]],
 
    [r'Why don\'?t you ([^\?]*)\??',
     ["Do you really think I don't {0}?",
      "Perhaps eventually I will {0}.",
      "Do you really want me to {0}?"]],
 
    [r'Why can\'?t I ([^\?]*)\??',
     ["Do you think you should be able to {0}?",
      "If you could {0}, what would you do?",
      "I don't know -- why can't you {0}?",
      "Have you really tried?"]],
 
    [r'I can\'?t (.*)',
     ["How do you know you can't {0}?",
      "Perhaps you could {0} if you tried.",
      "What would it take for you to {0}?"]],
 
    [r'I am (.*)',
     ["Did you come to me because you are {0}?",
      "How long have you been {0}?",
      "How do you feel about being {0}?"]],
 
    [r'I\'?m (.*)',
     ["How does being {0} make you feel?",
      "Do you enjoy being {0}?",
      "Why do you tell me you're {0}?",
      "Why do you think you're {0}?"]],
 
    [r'Are you ([^\?]*)\??',
     ["Why does it matter whether I am {0}?",
      "Would you prefer it if I were not {0}?",
      "Perhaps you believe I am {0}.",
      "I may be {0} -- what do you think?"]],
 
    [r'What (.*)',
     ["Why do you ask?",
      "How would an answer to that help you?",
      "What do you think?"]],
 
    [r'How (.*)',
     ["How do you suppose?",
      "Perhaps you can answer your own question.",
      "What is it you're really asking?"]],
 
    [r'Because (.*)',
     ["Is that the real reason?",
      "What other reasons come to mind?",
      "Does that reason apply to anything else?",
      "If {0}, what else must be true?"]],
 
    [r'(.*) sorry (.*)',
     ["There are many times when no apology is needed.",
      "What feelings do you have when you apologize?"]],
 
    [r'Hello(.*)',
     ["Hello... I'm glad you could drop by today.",
      "Hi there... how are you today?",
      "Hello, how are you feeling today?"]],
 
    [r'I think (.*)',
     ["Do you doubt {0}?",
      "Do you really think so?",
      "But you're not sure {0}?"]],
 
    [r'(.*) friend (.*)',
     ["Tell me more about your friends.",
      "When you think of a friend, what comes to mind?",
      "Why don't you tell me about a childhood friend?"]],
 
    [r'Yes',
     ["You seem quite sure.",
      "OK, but can you elaborate a bit?"]],
 
    [r'(.*) computer(.*)',
     ["Are you really talking about me?",
      "Does it seem strange to talk to a computer?",
      "How do computers make you feel?",
      "Do you feel threatened by computers?"]],
 
    [r'Is it (.*)',
     ["Do you think it is {0}?",
      "Perhaps it's {0} -- what do you think?",
      "If it were {0}, what would you do?",
      "It could well be that {0}."]],
 
    [r'It is (.*)',
     ["You seem very certain.",
      "If I told you that it probably isn't {0}, what would you feel?"]],
 
    [r'Can you ([^\?]*)\??',
     ["What makes you think I can't {0}?",
      "If I could {0}, then what?",
      "Why do you ask if I can {0}?"]],
 
    [r'Can I ([^\?]*)\??',
     ["Perhaps you don't want to {0}.",
      "Do you want to be able to {0}?",
      "If you could {0}, would you?"]],
 
    [r'You are (.*)',
     ["Why do you think I am {0}?",
      "Does it please you to think that I'm {0}?",
      "Perhaps you would like me to be {0}.",
      "Perhaps you're really talking about yourself?"]],
 
    [r'You\'?re (.*)',
     ["Why do you say I am {0}?",
      "Why do you think I am {0}?",
      "Are we talking about you, or me?"]],
 
    [r'I don\'?t (.*)',
     ["Don't you really {0}?",
      "Why don't you {0}?",
      "Do you want to {0}?"]],
 
    [r'I feel (.*)',
     ["Good, tell me more about these feelings.",
      "Do you often feel {0}?",
      "When do you usually feel {0}?",
      "When you feel {0}, what do you do?"]],
 
    [r'I have (.*)',
     ["Why do you tell me that you've {0}?",
      "Have you really {0}?",
      "Now that you have {0}, what will you do next?"]],
 
    [r'I would (.*)',
     ["Could you explain why you would {0}?",
      "Why would you {0}?",
      "Who else knows that you would {0}?"]],
 
    [r'Is there (.*)',
     ["Do you think there is {0}?",
      "It's likely that there is {0}.",
      "Would you like there to be {0}?"]],
 
    [r'My (.*)',
     ["I see, your {0}.",
      "Why do you say that your {0}?",
      "When your {0}, how do you feel?"]],
 
    [r'You (.*)',
     ["We should be discussing you, not me.",
      "Why do you say that about me?",
      "Why do you care whether I {0}?"]],
 
    [r'Why (.*)',
     ["Why don't you tell me the reason why {0}?",
      "Why do you think {0}?"]],
 
    [r'I want (.*)',
     ["What would it mean to you if you got {0}?",
      "Why do you want {0}?",
      "What would you do if you got {0}?",
      "If you got {0}, then what would you do?"]],
 
    [r'(.*) mother(.*)',
     ["Tell me more about your mother.",
      "What was your relationship with your mother like?",
      "How do you feel about your mother?",
      "How does this relate to your feelings today?",
      "Good family relations are important."]],
 
    [r'(.*) father(.*)',
     ["Tell me more about your father.",
      "How did your father make you feel?",
      "How do you feel about your father?",
      "Does your relationship with your father relate to your feelings today?",
      "Do you have trouble showing affection with your family?"]],
 
    [r'(.*) child(.*)',
     ["Did you have close friends as a child?",
      "What is your favorite childhood memory?",
      "Do you remember any dreams or nightmares from childhood?",
      "Did the other children sometimes tease you?",
      "How do you think your childhood experiences relate to your feelings today?"]],
 
    [r'(.*)\?',
     ["Why do you ask that?",
      "Please consider whether you can answer your own question.",
      "Perhaps the answer lies within yourself?",
      "Why don't you tell me?"]],
 
    [r'quit',
     ["Thank you for talking with me.",
      "Good-bye.",
      "Thank you, that will be $150.  Have a good day!"]],
 
    [r'(.*)',
     ["Please tell me more.",
      "Let's change focus a bit... Tell me about your family.",
      "Can you elaborate on that?",
      "Why do you say that {0}?",
      "I see.",
      "Very interesting.",
      "{0}.",
      "I see.  And what does that tell you?",
      "How does that make you feel?",
      "How do you feel when you say that?"]]
]
 
 
def reflect(fragment):
    tokens = fragment.lower().split()
    for i, token in enumerate(tokens):
        if token in reflections:
            tokens[i] = reflections[token]
    return ' '.join(tokens)
 
 
def analyze(statement):
    for pattern, responses in psychobabble:
        match = re.match(pattern, statement.rstrip(".!"))
        if match:
            response = random.choice(responses)
            return response.format(*[reflect(g) for g in match.groups()])
 


eliza.py
import re
import random

#MIT Eliza AI from several decades ago. Uses reflection to respond.
 
reflections = {
    "am": "are",
    "was": "were",
    "i": "you",
    "i'd": "you would",
    "i've": "you have",
    "i'll": "you will",
    "my": "your",
    "are": "am",
    "you've": "I have",
    "you'll": "I will",
    "your": "my",
    "yours": "mine",
    "you": "me",
    "me": "you"
}
 
psychobabble = [
    [r'I need (.*)',
     ["Why do you need {0}?",
      "Would it really help you to get {0}?",
      "Are you sure you need {0}?"]],
 
    [r'Why don\'?t you ([^\?]*)\??',
     ["Do you really think I don't {0}?",
      "Perhaps eventually I will {0}.",
      "Do you really want me to {0}?"]],
 
    [r'Why can\'?t I ([^\?]*)\??',
     ["Do you think you should be able to {0}?",
      "If you could {0}, what would you do?",
      "I don't know -- why can't you {0}?",
      "Have you really tried?"]],
 
    [r'I can\'?t (.*)',
     ["How do you know you can't {0}?",
      "Perhaps you could {0} if you tried.",
      "What would it take for you to {0}?"]],
 
    [r'I am (.*)',
     ["Did you come to me because you are {0}?",
      "How long have you been {0}?",
      "How do you feel about being {0}?"]],
 
    [r'I\'?m (.*)',
     ["How does being {0} make you feel?",
      "Do you enjoy being {0}?",
      "Why do you tell me you're {0}?",
      "Why do you think you're {0}?"]],
 
    [r'Are you ([^\?]*)\??',
     ["Why does it matter whether I am {0}?",
      "Would you prefer it if I were not {0}?",
      "Perhaps you believe I am {0}.",
      "I may be {0} -- what do you think?"]],
 
    [r'What (.*)',
     ["Why do you ask?",
      "How would an answer to that help you?",
      "What do you think?"]],
 
    [r'How (.*)',
     ["How do you suppose?",
      "Perhaps you can answer your own question.",
      "What is it you're really asking?"]],
 
    [r'Because (.*)',
     ["Is that the real reason?",
      "What other reasons come to mind?",
      "Does that reason apply to anything else?",
      "If {0}, what else must be true?"]],
 
    [r'(.*) sorry (.*)',
     ["There are many times when no apology is needed.",
      "What feelings do you have when you apologize?"]],
 
    [r'Hello(.*)',
     ["Hello... I'm glad you could drop by today.",
      "Hi there... how are you today?",
      "Hello, how are you feeling today?"]],
 
    [r'I think (.*)',
     ["Do you doubt {0}?",
      "Do you really think so?",
      "But you're not sure {0}?"]],
 
    [r'(.*) friend (.*)',
     ["Tell me more about your friends.",
      "When you think of a friend, what comes to mind?",
      "Why don't you tell me about a childhood friend?"]],
 
    [r'Yes',
     ["You seem quite sure.",
      "OK, but can you elaborate a bit?"]],
 
    [r'(.*) computer(.*)',
     ["Are you really talking about me?",
      "Does it seem strange to talk to a computer?",
      "How do computers make you feel?",
      "Do you feel threatened by computers?"]],
 
    [r'Is it (.*)',
     ["Do you think it is {0}?",
      "Perhaps it's {0} -- what do you think?",
      "If it were {0}, what would you do?",
      "It could well be that {0}."]],
 
    [r'It is (.*)',
     ["You seem very certain.",
      "If I told you that it probably isn't {0}, what would you feel?"]],
 
    [r'Can you ([^\?]*)\??',
     ["What makes you think I can't {0}?",
      "If I could {0}, then what?",
      "Why do you ask if I can {0}?"]],
 
    [r'Can I ([^\?]*)\??',
     ["Perhaps you don't want to {0}.",
      "Do you want to be able to {0}?",
      "If you could {0}, would you?"]],
 
    [r'You are (.*)',
     ["Why do you think I am {0}?",
      "Does it please you to think that I'm {0}?",
      "Perhaps you would like me to be {0}.",
      "Perhaps you're really talking about yourself?"]],
 
    [r'You\'?re (.*)',
     ["Why do you say I am {0}?",
      "Why do you think I am {0}?",
      "Are we talking about you, or me?"]],
 
    [r'I don\'?t (.*)',
     ["Don't you really {0}?",
      "Why don't you {0}?",
      "Do you want to {0}?"]],
 
    [r'I feel (.*)',
     ["Good, tell me more about these feelings.",
      "Do you often feel {0}?",
      "When do you usually feel {0}?",
      "When you feel {0}, what do you do?"]],
 
    [r'I have (.*)',
     ["Why do you tell me that you've {0}?",
      "Have you really {0}?",
      "Now that you have {0}, what will you do next?"]],
 
    [r'I would (.*)',
     ["Could you explain why you would {0}?",
      "Why would you {0}?",
      "Who else knows that you would {0}?"]],
 
    [r'Is there (.*)',
     ["Do you think there is {0}?",
      "It's likely that there is {0}.",
      "Would you like there to be {0}?"]],
 
    [r'My (.*)',
     ["I see, your {0}.",
      "Why do you say that your {0}?",
      "When your {0}, how do you feel?"]],
 
    [r'You (.*)',
     ["We should be discussing you, not me.",
      "Why do you say that about me?",
      "Why do you care whether I {0}?"]],
 
    [r'Why (.*)',
     ["Why don't you tell me the reason why {0}?",
      "Why do you think {0}?"]],
 
    [r'I want (.*)',
     ["What would it mean to you if you got {0}?",
      "Why do you want {0}?",
      "What would you do if you got {0}?",
      "If you got {0}, then what would you do?"]],
 
    [r'(.*) mother(.*)',
     ["Tell me more about your mother.",
      "What was your relationship with your mother like?",
      "How do you feel about your mother?",
      "How does this relate to your feelings today?",
      "Good family relations are important."]],
 
    [r'(.*) father(.*)',
     ["Tell me more about your father.",
      "How did your father make you feel?",
      "How do you feel about your father?",
      "Does your relationship with your father relate to your feelings today?",
      "Do you have trouble showing affection with your family?"]],
 
    [r'(.*) child(.*)',
     ["Did you have close friends as a child?",
      "What is your favorite childhood memory?",
      "Do you remember any dreams or nightmares from childhood?",
      "Did the other children sometimes tease you?",
      "How do you think your childhood experiences relate to your feelings today?"]],
 
    [r'(.*)\?',
     ["Why do you ask that?",
      "Please consider whether you can answer your own question.",
      "Perhaps the answer lies within yourself?",
      "Why don't you tell me?"]],
 
    [r'quit',
     ["Thank you for talking with me.",
      "Good-bye.",
      "Thank you, that will be $150.  Have a good day!"]],
 
    [r'(.*)',
     ["Please tell me more.",
      "Let's change focus a bit... Tell me about your family.",
      "Can you elaborate on that?",
      "Why do you say that {0}?",
      "I see.",
      "Very interesting.",
      "{0}.",
      "I see.  And what does that tell you?",
      "How does that make you feel?",
      "How do you feel when you say that?"]]
]
 
 
def reflect(fragment):
    tokens = fragment.lower().split()
    for i, token in enumerate(tokens):
        if token in reflections:
            tokens[i] = reflections[token]
    return ' '.join(tokens)
 
 
def analyze(statement):
    for pattern, responses in psychobabble:
        match = re.match(pattern, statement.rstrip(".!"))
        if match:
            response = random.choice(responses)
            return response.format(*[reflect(g) for g in match.groups()])
 


bstest6_3.py
# -*- coding: utf-8 -*-
from bs4 import BeautifulSoup
from random import randint
import lxml

import urllib
import time


class foodSites:


    ingstr = ""
    linklist = ""

    def ingredList(self, site):
        htmlpg = urllib.urlopen("http://allrecipes.com"+site)
        soup = BeautifulSoup(htmlpg,"lxml")
        ingr = soup.findAll('li', class_= "checkList__line")
        fin = ""
        for ele in ingr:
            if ele.span.get_text()!="" and ele.span.get_text()!="Add all ingredients to list":
                fin = fin + ele.span.get_text()+ "\n"
        return str(fin)

    def findLink(self, theLink):
        htmlpg = urllib.urlopen(theLink)
        soup = BeautifulSoup(htmlpg,"lxml")
        print(soup.title.get_text())
        recip_list = soup.findAll('article', class_="grid-col--fixed-tiles")
        listr = []
        for element in recip_list:
                if element.a is not None:
                        if "/recipe/" in element.a["href"]:
                                print(element.a["href"])
                                listr.append(element.a["href"])
        num = randint(0, len(listr)-1)
        print("\n")
        return listr[num]


    def initList(self):
        global ingstr
        global linklist
        link1 = "http://allrecipes.com/recipes/446/main-dish/beef/"
        #link2 = "http://allrecipes.com/recipes/16954/main-dish/chicken/"
        #link3 = "http://allrecipes.com/recipes/17245/main-dish/pasta/"
        
        chosen = self.findLink(link1)
        ingstr = "Ingredients: \n"+self.ingredList(chosen)
        linklist = "http://allrecipes.com" + chosen + "\n"

        #chosen = self.findLink(link2)
        #ingstr = ingstr + self.ingredList(chosen)
        #linklist = linklist + "http://allrecipes.com" + chosen + "\n"

        #chosen = self.findLink(link3)
        #ingstr = ingstr + self.ingredList(chosen)
        #linklist = linklist + "http://allrecipes.com" + chosen + "\n"

        print(ingstr)

    def getIngred(self):
        global ingstr
        return ingstr

    def getSites(self):
        global linklist
        return linklist


pyjbot.py
from flask import Flask, request
import json
import requests
import urllib

#for getting logs
import sys
import logging

from wit import Wit
import BeautifulSoup #not being used
import time

#import aiml

#grocery list class
from bstest6_3 import foodSites
import eliza

import os

app = Flask(__name__)

#alice
#kernel = aiml.Kernel()
#kernel.learn("std-startup.xml")
#kernel.respond("load aiml b")

#logs errors for heroku 
app.logger.addHandler(logging.StreamHandler(sys.stdout))
app.logger.setLevel(logging.ERROR)

site = ''
chatAl = False
loc = False
lat = ''
lon = ''

# This needs to be filled with the Page Access Token that will be provided
# by the Facebook App that will be created.
PAT = str(os.environ.get('FBKey',3))

#verify
@app.route('/', methods=['GET'])
def handle_verification():
  print "Handling Verification."
  if request.args.get('hub.verify_token', '') == 'my_voice_is_my_password_verify_me':
    print "Verification successful!"
    return request.args.get('hub.challenge', '')
  else:
    print "Verification failed!"
    return 'Error, wrong validation token'


#messaging
@app.route('/', methods=['POST'])
def handle_messages():
  #variables
  global site
  global chatAl
  global loc
  global lat 
  global lon
  
  context0 = {}
  count = 0
  
  print "Handling Messages"
  payload = request.get_data()
  print payload
  data = json.loads(payload)
  msgev = data["entry"][0]["messaging"]
  for event in msgev:
    if "message" in event:
      if "attachments" in event["message"]:
        for atta in event["message"]["attachments"]:
          if "payload" in atta:
            if "coordinates" in atta["payload"]:
              print(atta["payload"]["coordinates"])
              lat = atta["payload"]["coordinates"]["lat"]
              lon = atta["payload"]["coordinates"]["long"]
              f =  open("geo.txt", "w")
              f.write(str(lat)+","+str(lon))
              send_message(PAT, event["sender"]["id"], str(atta["payload"]["coordinates"]))
              send_message(PAT,event["sender"]["id"], "Coordinates Recieved")
              loc = True
              return "ok"
  
  #checks if chat option is on or not
  for sender, message in messaging_events(payload):
    
    #chatting with eliza ai
    if chatAl:
      #end chat
      if message == "bye":
        chatAl =False
        
      #chat and get response
      m1 = eliza.analyze(message)
      print("Trying to send...")
      send_message(PAT, sender, m1)
      print("Probably sent")
    
    #not chatting. Use Wit.AI
    else:
      
      print "Incoming from %s: %s" % (sender, message)
      print type(message)
      resp = client.message(message) #get response from wit.ai
      print(resp)
      
      #decides what type of response it is
      if u'entities' in resp:
        resp = resp[u'entities']
        if u'intent' in resp:
          resp = resp[u'intent']
        else:
          send_message(PAT, sender, "I'm sorry. I couldn't understand you. Please rephrase that.")
          return "ok"
      else:
        send_message(PAT, sender, "I'm sorry. I couldn't understand you. Please rephrase that.")
        return "ok"
        
      #if response is understood, get it
      resp = resp[0]
      print ("Response type is.... "+resp[u'value'])
      
      
      #id type of response and run correct method
      if u'value' in resp:
        
        #grocery response to get list of groceries. From imported class
        if resp[u'value'] == "grocery":
          message = get_cooking()
          while( len(message) > 300):
            msg2 = message[:300]
            message = message[300:]
            send_message(PAT, sender, msg2)
          send_message(PAT, sender, message)
          send_message(PAT, sender, site) 
          
            
        #get xkcd comic link, poorly implemented rn
        elif resp[u'value'] == "xkcd":
          message = "http://xkcd.com/"
          send_message(PAT, sender, message)
          
        #Location data to switch modes
        elif resp[u'value'] == "restaurants":
          restaurants = ''
          print("rest method")
          message = get_restaurants(sender)
          if message['status']== "OK":
            for place in message['results']:
              if 'permanently_closed' not in place or place['permanently_closed'] is False:
                #if 'opennow' in place and place['opennow'] is True:
                restaurant = "Name: "+str(place['name']) +"  Rating: "
                if 'rating' in place:
                  restaurant+= str(place['rating'])
                else:
                  restaurant+= "None"
                  
                send_message(PAT, sender, restaurant)
          else:
            send_message(PAT, sender, "Error")
          #message = str(get_restaurants(sender))
          #while( len(message) > 300):
          #  msg2 = message[:300]
          #  message = message[300:]
          #  send_message(PAT, sender, msg2)
          #send_message(PAT, sender, message)
          #else:
          #send_message(PAT, sender, "Not ok")
          
          #else:
           # send_message(PAT, sender, "Enter your location:")
          #loc = True
          time.sleep(6)

            
        #greetings response. Usually used to start up
        elif resp[u'value'] == "greetings":
          print("This resp greetings RIGHT HERE")
          resp = client.converse('my-user-session-42',message, context0)
          print("This resp greetings ")
          print (resp)
          
          while('msg' not in resp):
            resp = client.converse('my-user-session-42',message, context0)
            print ("This resp HERE ")
            print(resp)
            
          print("the msg is "+resp['msg'])
          message = str(resp["msg"])
          print("Trying to send...")
          send_message(PAT, sender, message)
          print("Probably sent")
        
          
        #talk sets chatAI to true and allows chat with eliza. Use "bye" to end  
        elif resp[u'value'] == "talk":
          chatAl =True
          send_message(PAT, sender, "Okay, what's up?")
        
        
        #not working atm
        elif resp[u'value'] == "weather":
          #resp = client.run_actions('my-user-session-42',textmsg, context0)
          print("This resp weather ")
          #print (resp)
          #while('foodList' not in resp):
          #    resp = client.run_actions('my-user-session-42',textmsg, context0)
          #    print ("This resp ")
          #    print(resp)
          #voice.send_sms(msg[u'from'],str(resp['forecast']))
          message = "weather"
          send_message(PAT, sender, message)
            
            
        #catch all for other intents  
        else:
          print("Else")
          resp = client.converse('my-user-session-42',message, context0)
          while('msg' not in resp and count <=10):
            resp = client.converse('my-user-session-42',message, context0)
            print ("This resp HERE ")
            print(resp)
              
          print("the msg is "+resp['msg'])
          message = str(resp["msg"])
          print("Trying to send...")
          send_message(PAT, sender, message)
  return "ok"

#Sorts messages
def messaging_events(payload):
  """Generate tuples of (sender_id, message_text) from the
  provided payload.
  """
  data = json.loads(payload)
  messaging_events = data["entry"][0]["messaging"]
  for event in messaging_events:
    if "message" in event and "text" in event["message"]:
      yield event["sender"]["id"], event["message"]["text"].encode('unicode_escape')
    else:
      yield event["sender"]["id"], "I can't echo this"


#Send the message. Limited to 320 char
def send_message(token, recipient, text):
  """Send the message text to recipient with id recipient.
  """

  r = requests.post("https://graph.facebook.com/v2.6/me/messages",
    params={"access_token": token},
    data=json.dumps({
      "recipient": {"id": recipient},
      "message": {"text": text.decode('unicode_escape')}
    }),
    headers={'Content-type': 'application/json'})
  if r.status_code != requests.codes.ok:
    print r.text

#wit ai method not implemented
def merge():
    return 0

#wit ai method not implemented (important probably)
def error():
    return 0

#wit ai method works
def say(id, dict, response):
    print ("id " +id)
    print (dict)
    print ("respo "+response)
    return response
    
#wit ai send method
def send(request, response):
    print(response['text'])

#wit ai function method
def my_action(request):
    print('Received from user...', request['text'])

#gotta use a weather module
def get_forecast(request):
    context = request['context']
    entities = request['entities']

    loc = first_entity_value(entities, 'location')
    if loc:
        context['forecast'] = 'sunny'
    else:
        context['missingLocation'] = True
        if context.get('forecast') is not None:
            del context['forecast']

    return context

#works fine so far. Can't run from wit.ai
def get_cooking():
    print("Inside grocery")
    global site
    context = ""
    cook = foodSites()
    cook.initList()
    context= cook.getIngred()
    site = cook.getSites()
    return context
    
    
def get_restaurants(sender):
  f = open("geo.txt", "r")
  Location = f.read()
  send_message(PAT, sender, str(Location))
  print("finished loc")
  loc_url = "https://maps.googleapis.com/maps/api/place/nearbysearch/json?location="+Location+"&rankby=distance&keyword=restaurant&key="+str(os.environ.get('GAPI',3))
  print("url done")
  resp = urllib.urlopen(loc_url)
  print("json read")
  data = resp.read()
  jData = json.loads(data)
  print("json loaded")
  return jData
  
  
#wit ai action list
actions = {
    'send': send,
    'getForecast': get_forecast,
    'say': say,
    'merge': merge,
    'error': error,
    "getCooking": get_cooking,
    'my_action': my_action,
}

#server access
client = Wit(access_token=str(os.environ.get('ATKey',3)), actions=actions)


if __name__ == '__main__':
  app.run()
